merge.py

```diff
@@ -1,98 +1,303 @@
+        work_queue: A Queue.Queue where new work items will be queued.
+    Returns:
+        Merger of type specified in configuration
+    Raises:
+        AttributeError: if passed an unsupported SCM type.
+    """
+    if config['scm_type'] == 'github':
+        return GitMerger(config, work_queue)
+    raise AttributeError(
+        'Unsupported SCM type: {}.'.format(config['scm_type']))
+
+
+class Merger(Thread):
+    """Merger is the base class for all mergers.
+    """
+    __metaclass__ = abc.ABCMeta
+
+    def __init__(self, config, work_queue):
+        self.config = config
+        l = logging.getLogger('{name}_merge_logger'.format(
+            name=self.config['name']))
+        log_fmt = '%(asctime)s - %(name)s - %(levelname)s - %(message)s'
+        f = logging.Formatter(log_fmt)
+        filename = '{name}_merger_log.txt'.format(name=config['name'])
+        h = logging.FileHandler(os.path.join('log', filename), mode='w')
+        h.setFormatter(f)
+        l.addHandler(h)
+        l.setLevel(logging.INFO)
+        self.main_logger = l
+        self.work_queue = work_queue
+        Thread.__init__(self)
+
+    @abc.abstractmethod
+    def run(self):
+        """run spins forever, merging things off of the work queue.
+        """
+        return
+
+
+class GitMerger(Merger):
+    """GitMerger merges Git pull requests.
+    """
+
+    def run(self):
+        """run spins forever, merging things off of the work queue.
+
+        Raises:
+            AttributeError: if work items are not of type github_helper.GithubPR
+        """
+        while True:
+            pr = self.work_queue.get()
+            pr_num = pr.get_num()
+            self.main_logger.info(
+                'Starting work on PR#{pr_num}.'.format(pr_num=pr_num))
+            self.main_logger.info('{remaining} work items remaining.'.format(
+                remaining=self.work_queue.qsize()))
+
+            m_l = self.get_logger()
+            m_l.info('Starting merge process for #{}.'.format(pr_num))
+
+            tmp_dir = TMP_DIR_FMT.format(dir='{}-{}'.format(
+                self.config['repository'], pr_num))
+            try:
+                _set_up(tmp_dir)
+            except AssertionError:
+                pr.post_error('Setup of temp directory failed, try again.')
+                continue
+            if self.merge_git_pr(pr, tmp_dir, m_l):
+                m_l.info('Merge concluded satisfactorily. Moving on.')
+                self.main_logger.info(
+                    'PR#{num} processing done.'.format(num=pr_num))
+            else:
+                m_l.info('Merge did not conclude satisfactorily (reporting to'
+                         ' github failed). Adding PR back to queue to be tried'
+                         ' again.')
+                self.main_logger.info(
+                    'PR#{num} processing failed.'.format(num=pr_num))
+                self.work_queue.put(pr)
+            _clean_up(tmp_dir)
+
+    def get_logger(self, pr_num):
```

Comment:
This shouldn't be part of the GitMerger class. It seems like there should be a logging utility added that has utilities for things like `GetPrLogger`. Is it possible to make the logging changes in a separate PR? 

I'd suggest having a MergebotLogger class, and a MergebotPrLogger class. The MergebotLogger has a method like `forPR(pr_num)`. The main method creates the MergebotLogger and thats what everything gets. The GitMerger class uses the forPR to specialize it and carries that around as appropriate. And it can have methods for logging just to the main logger, logging to both, and logging to just the PR log (or maybe different log levels are redirected differently).